libraries/BLE/src/BLECallbacks.cpp

```diff
@@ -0,0 +1,207 @@
+}
+
+ssize_t profile_longwrite_process(struct bt_conn *conn,
+                                     const struct bt_gatt_attr *attr,
+                                     const void *buf, uint16_t len,
+                                     uint16_t offset)
+{
+    BLECharacteristicImp *blecharacteritic = (BLECharacteristicImp*)attr->user_data;
+    
+    blecharacteritic->setBuffer((const uint8_t *) buf, len, offset);
+    
+    return len;
+}
+
+int profile_longflush_process(struct bt_conn *conn,
+                              const struct bt_gatt_attr *attr, 
+                              uint8_t flags)
+{
+    BLECharacteristicImp *blecharacteritic = (BLECharacteristicImp*)attr->user_data;
+
+    switch (flags)
+    {
+    case BT_GATT_FLUSH_DISCARD:
+        /* Discard buffer reseting it back with data */
+        blecharacteritic->discardBuffer();
+        return 0;
+    case BT_GATT_FLUSH_SYNC:
+        /* Sync buffer to data */
+        blecharacteritic->syncupBuffer2Value();
+        return 0;
+    }
+
+    return -EINVAL;
+}
+
+
+// GATT client only
+uint8_t profile_notify_process (bt_conn_t *conn,
+                                bt_gatt_subscribe_params_t *params,
+                                const void *data, uint16_t length)
+{
+    //BLEPeripheralHelper* peripheral = BLECentralRole::instance()->peripheral(conn);// Find peripheral by bt_conn
+    //BLEAttribute* notifyatt = peripheral->attribute(params); // Find attribute by params
+    BLECharacteristicImp* chrc = NULL;
+    BLEDevice bleDevice(bt_conn_get_dst(conn));
+    chrc = BLEProfileManager::instance()->characteristic(bleDevice, params->value_handle);
+    
+    //assert(notifyatt->type() == BLETypeCharacteristic);
+    pr_debug(LOG_MODULE_APP, "%s1", __FUNCTION__);
+    if (NULL != chrc)
+    {
+        chrc->setValue((const unsigned char *)data, length);
+    }
+    return BT_GATT_ITER_CONTINUE;
+}
+
+// GATT client only
+uint8_t profile_discover_process(bt_conn_t *conn,
+                                 const bt_gatt_attr_t *attr,
+                                 bt_gatt_discover_params_t *params)
+{
+    uint8_t ret = BT_GATT_ITER_STOP;
+    pr_debug(LOG_MODULE_BLE, "%s-%d", __FUNCTION__, __LINE__);
+    ret = BLEProfileManager::instance()->discoverResponseProc(conn, attr, params);
+    pr_debug(LOG_MODULE_BLE, "%s-%d", __FUNCTION__, __LINE__);
+    return ret;
+}
+
+// GATT Client only
+uint8_t profile_read_rsp_process(bt_conn_t *conn, 
+                                 int err,
+                                 bt_gatt_read_params_t *params,
+                                 const void *data, 
+                                 uint16_t length)
+{
+    if (NULL == data)
+    {
+        return BT_GATT_ITER_STOP;
+    }
+    BLECharacteristicImp *chrc = NULL;
+    BLEDevice bleDevice(bt_conn_get_dst(conn));
+    
+    // Get characteristic by handle params->single.handle
+    chrc = BLEProfileManager::instance()->characteristic(bleDevice, params->single.handle);
```

Comment:
Remember, every time you call `BLEProfileManager::instance()`, it creates a new `BLEProfileManager` on the (very small) heap. Running out of heap space is very possible with < 8k available. Always check for NULL-ness.